### ISSUE TITLE
Difference between "loss" and "neg_log_perplexity" in "perplexity_eval" mode

### ISSUE THREAD

**Issue by arglog** [google-research/text-to-text-transfer-transformer#356]
Hi

In the perplexity_eval model (using `perplexity_eval.gin`), there are two metrics being recorded: `loss` ([L761](https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/utils.py#L761)) and `neg_log_perplexity` ([L798](https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/utils.py#L798)). 

I found these two losses are essentially cross entropy loss: `loss` was computed from [here](https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/transformer.py#L777) and `neg_log_perplexity` was based on values computed [here](https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/utils.py#L767).

My question is: what's the difference between `loss` and `neg_log_perplexity`? In my experiments I found that `loss` is not equal to `-1 * neg_log_perplexity`. What's the reason that makes the two differenct?

Thanks!

**Comment by arglog** [comment]
One difference I found is that in `logits_and_loss` the cross entropy was normalized by the [length](https://github.com/tensorflow/mesh/blob/a3c05f705641dfe144f70b7b5230db4933ce8ca9/mesh_tensorflow/transformer/transformer.py#L788) of `targets`, while in `neg_log_perplexity` the cross entropy was normalized by the [number of non-zero tokens](https://github.com/tensorflow/mesh/blob/a3c05f705641dfe144f70b7b5230db4933ce8ca9/mesh_tensorflow/transformer/utils.py#L777) in `targets`.
